<center><h1> Case Study 1</h1></center>
<center><h3> Week 1 (out of 5)</h3></center>

**Author of this file:**
1. Will Schmidt (will.schmidt@emory.edu)
 
**Data Source**: W.C. Hunter and M.B. Walker (1996), [“*The Cultural Affinity Hypothesis and Mortgage Lending Decisions*,”](https://link.springer.com/article/10.1007/BF00174551) Journal of Real Estate Finance and Economics 13, 57-70.
 
**Book**: [Introductory Econometrics: A Modern Approach](https://economics.ut.ac.ir/documents/3030266/14100645/Jeffrey_M._Wooldridge_Introductory_Econometrics_A_Modern_Approach__2012.pdf) by Jeffrey Wooldridge

**Data Description**: ```http://fmwww.bc.edu/ec-p/data/wooldridge/loanapp.dta```

```
  Obs:  1989

  1. occ                       occupancy
  2. loanamt                   loan amt in thousands
  3. action                    type of action taken
  4. msa                       msa number of property
  5. suffolk                   =1 if property in Suffolk County
  6. race                      race of applicant
  7. gender                    gender of applicant
  8. appinc                    applicant income, $1000s
  9. typur                     type of purchaser of loan
 10. unit                      number of units in property
 11. married                   =1 if applicant married
 12. dep                       number of dependents
 13. emp                       years employed in line of work
 14. yjob                      years at this job
 15. self                      self-employment dummy
 16. atotinc                   total monthly income
 17. cototinc                  coapp total monthly income
 18. hexp                      propose housing expense
 19. price                     purchase price
 20. other                     other financing, $1000s
 21. liq                       liquid assets
 22. rep                       no. of credit reports
 23. gdlin                     credit history meets guidelines
 24. lines                     no. of credit lines on reports
 25. mortg                     credit history on mortgage paym
 26. cons                      credit history on consumer stuf
 27. pubrec                    =1 if filed bankruptcy
 28. hrat                      housing exp, % total inccome
 29. obrat                     other oblgs,  % total income
 30. fixadj                    fixed or adjustable rate?
 31. term                      term of loan in months
 32. apr                       appraised value
 33. prop                      type of property
 34. inss                      PMI sought
 35. inson                     PMI approved
 36. gift                      gift as down payment
 37. cosign                    is there a cosigner
 38. unver                     unverifiable info
 39. review                    number of times reviewed
 40. netw                      net worth
 41. unem                      unemployment rate by industry
 42. min30                     =1 if minority pop. > 30%
 43. bd                        =1 if boarded-up val > MSA med
 44. mi                        =1 if tract inc > MSA median
 45. old                       =1 if applic age > MSA median
 46. vr                        =1 if tract vac rte > MSA med
 47. sch                       =1 if > 12 years schooling
 48. black                     =1 if applicant black
 49. hispan                    =1 if applicant Hispanic
 50. male                      =1 if applicant male
 51. reject                    =1 if action == 3
 52. approve                   =1 if action == 1 or 2
 53. mortno                    no mortgage history
 54. mortperf                  no late mort. payments
 55. mortlat1                  one or two late payments
 56. mortlat2                  > 2 late payments
 57. chist                     =0 if accnts deliq. >= 60 days
 58. multi                     =1 if two or more units
 59. loanprc                   amt/price
 60. thick                     =1 if rep > 2
 61. white                     =1 if applicant white
 62. obwhte                    obrat*awhite
 ```

<center><h1> Questions</h1></center>

1. Using the 62 variables in the data set create a new data frame containing _all_ the 26 explanatory variables described in Table 1 (page 61) as well as the outcome variable `action`.

In [41]:
#import necessary data
import pandas as pd
df = pd.read_stata('http://fmwww.bc.edu/ec-p/data/wooldridge/loanapp.dta')

In [42]:
# match varible names in the paper and dataset
# according to the definition in Table 1, pub = pubrec
# however, based on the summary statistics in Table 2, pub = 1 - pubrec
# Therefore, I adjust pub = 1 - pubrec. Similarly for Morhist and location 

# uemp = unem
# ms = married
# lnpr = loanprc
# thk = thick
# asex = male
# arac = white
df['pub'] = 1 - df['pubrec']
df['mhist'] = 1 - df['mortlat2']
df['loc'] = 1 - df['vr']

var = ['approve', 'hrat', 'obrat', 'mhist', 'pub', 'self', 'chist', 'unem', 'multi', 'cosign', 'married',
      'loanprc', 'dep', 'sch', 'thick', 'white', 'male', 'loc']

#create our interaction terms
tmp = ['white', 'male', 'thick', 'loc']
for x in tmp:
    df['sch'+x] = df['sch']*df[x]
    var.append('sch' + x)
for x in tmp:
    df['chist'+x] = df['chist']*df[x]
    var.append('chist' + x)

# obrac = obwhte
var.append('obwhte') # var is now a complete list of names of all variables we need in df
df_new = df[var]

# change variable names as in the paper
varname = ['approve', 'hrat', 'obrat', 'mhist', 'pub', 'self', 'chist', 'uemp', 'multi', 'cosign', 'ms', 
           'lnpr', 'dep', 'sch', 'thk', 'arac', 'asex', 'loc', 'schrac','achsex','schthk','scheloc','chrac',
           'chsex','shthk','chloc','obrac']
varname = list(map(lambda x:x.upper(), varname)) # upper case
df_new.columns = varname

# display the shape and columns names of df_new to indicate it is the data frame we want
print(f'The shape of the data frame is {df_new.shape}.')
print(f'The columns are {list(df_new.columns)}.')

The shape of the data frame is (1989, 27).
The columns are ['APPROVE', 'HRAT', 'OBRAT', 'MHIST', 'PUB', 'SELF', 'CHIST', 'UEMP', 'MULTI', 'COSIGN', 'MS', 'LNPR', 'DEP', 'SCH', 'THK', 'ARAC', 'ASEX', 'LOC', 'SCHRAC', 'ACHSEX', 'SCHTHK', 'SCHELOC', 'CHRAC', 'CHSEX', 'SHTHK', 'CHLOC', 'OBRAC'].


2. After reading the original article try to _replicate_ Table 2 (page 62) using the 1989 observations. **Note**: The numbers will not be _exactly_ the same because the provided data set is missing 2 observations from the original 1,991 observations used in the study.

In [43]:
# calculate stats and transpose
t2 = df_new.describe().loc[['mean','std','min','max']].transpose().iloc[1:]
t2.reset_index(inplace = True)

# reset column names 
colname = ['Variable Name', 'Mean', 'Standard Deviation', 'Minimum', 'Maximum']
t2.columns = colname

# make index blank
t2.index = ['']*t2.shape[0]

t2 = t2.round(3)
print('Table 2. Summary statistics for variables in model. ')
t2

Table 2. Summary statistics for variables in model. 


Variable Name    Mean  Standard Deviation  Minimum  Maximum
          HRAT  24.791               7.119    1.000   72.000
         OBRAT  32.389               8.263    0.000   95.000
         MHIST   0.989               0.102    0.000    1.000
           PUB   0.931               0.253    0.000    1.000
          SELF   0.129               0.336    0.000    1.000
         CHIST   0.838               0.369    0.000    1.000
          UEMP   3.882               2.164    1.800   10.600
         MULTI   0.086               0.281    0.000    1.000
        COSIGN   0.029               0.167    0.000    1.000
            MS   0.659               0.474    0.000    1.000
          LNPR   0.771               0.189    0.021    2.571
           DEP   0.771               1.104    0.000    8.000
           SCH   0.772               0.420    0.000    1.000
           THK   0.105               0.307    0.000    1.000
          ARAC   0.845               0.362    0.000    1.000
          ASEX   0.813               0.390    0.000    1.000
           LOC   0.590               0.492    0.000    1.000
        SCHRAC   0.668               0.471    0.000    1.000
        ACHSEX   0.620               0.485    0.000    1.000
        SCHTHK   0.083               0.276    0.000    1.000
       SCHELOC   0.465               0.499    0.000    1.000
         CHRAC   0.727               0.446    0.000    1.000
         CHSEX   0.680               0.467    0.000    1.000
         SHTHK   0.083               0.276    0.000    1.000
         CHLOC   0.497               0.500    0.000    1.000
         OBRAC  27.061              13.828    0.000   95.000

3. Replicate the _Estimated Coefficient_ column of Table 3 (page 63). You are also expected to print the _Variable Name_ column. **Note:** You are strongly advised to use the [Logit](https://www.statsmodels.org/stable/generated/statsmodels.discrete.discrete_model.Logit.html#statsmodels.discrete.discrete_model.Logit) command from the `statsmodels` library to answer this question. 

In [44]:
import statsmodels.api as sm
df_new = df_new.dropna()
X = df_new.drop('APPROVE', axis = 1)
X = sm.add_constant(X)

# according to table3 in the paper, pub = pubrec
X['PUB'] = 1-X['PUB']
y = df_new[['APPROVE']]
reg = sm.Logit(y, X).fit()

coef = pd.DataFrame(reg.summary().tables[1]).iloc[2:, 1]
t3 = pd.DataFrame({
    'Variable Name': varname[1:],
    'Estimated Coefficient': coef
})

# make index blank
t3.index = ['']*t3.shape[0]

print('Table 3. Logit results on mortgage acceptances\n(dependent variable = 1 if mortgage was accepted, 0 if mortgage was denied). ')
t3

Optimization terminated successfully.
         Current function value: 0.299449
         Iterations 7
Table 3. Logit results on mortgage acceptances
(dependent variable = 1 if mortgage was accepted, 0 if mortgage was denied). 


Variable Name Estimated Coefficient
          HRAT                0.0127
         OBRAT               -0.0662
         MHIST                0.7371
           PUB               -1.3870
          SELF               -0.5550
         CHIST                0.9179
          UEMP               -0.0530
         MULTI               -0.5198
        COSIGN                0.3572
            MS                0.4683
          LNPR               -1.8652
           DEP               -0.0837
           SCH               -0.8557
           THK               -0.4967
          ARAC                0.5939
          ASEX               -0.7405
           LOC               -0.3427
        SCHRAC               -0.3766
        ACHSEX                0.9299
        SCHTHK                0.7567
       SCHELOC                0.4327
         CHRAC               -0.0590
         CHSEX               -0.1104
         SHTHK               -0.0814
         CHLOC                0.4945
         OBRAC                0.0185

4. After partitioning the original 1,989 observations into a training data set (80%) and validation set (20%) using a `random_state` equal to 42, proceed to train the model you fit in 3 above using the `LogisticRegression` function from the `sklearn` library with the training data set and then print the proportion of true predictions for your validation set.

In [45]:
from sklearn.model_selection import train_test_split
X = X.drop('const', axis = 1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [46]:
from sklearn.linear_model import LogisticRegression
import numpy as np
logit1 = LogisticRegression(fit_intercept=True,max_iter=1000,solver='lbfgs',penalty='none').fit(X_train, y_train.values.ravel())
print('The proportion of true predictions for the validation set is', logit1.score(X_test, y_test))

The proportion of true predictions for the validation set is 0.8829516539440203


5. Can you _build_ a better model? Using the *Elastic Net* discussed in class, find an alternative model specification that having been trained using the 80% of the original sample has a *higher* proportion of true predictions for the validation set that your found in 4 above. 

For the purpose of choosing a model, I will try lasso and ridge separating and continue with one method. First I include all the variables and their interaction terms in the model. 

In [47]:
# I interact all the explanatory variables with race, sex and thickness of file to find variables that result in a higher true predictions than 88.2952%.
# I find that 'LOC' lowers true predictions, so I exclude it from interaction.
# I also discover that 'UEMP','SCH' and 'OBRAT' increase the proportion significantly while 'MHIST' decreases the proportion. Other variables do not significantly affect the proportion.
# Thus, I remove 'MHIST' and 'LOC' from interaction terms and add 'UEMP' and 'OBRAT' as new interaction terms.

# Create a new dataframe
df1 = df_new[['APPROVE','HRAT','OBRAT','MHIST','PUB','SELF','CHIST','UEMP','MULTI','COSIGN','MS','LNPR','DEP','SCH','THK','ARAC','ASEX','LOC','SCHRAC','ACHSEX','SCHTHK','SCHELOC']].copy()

# Create new interaction variables
tmp1 = ['ARAC','ASEX','THK']
for x in tmp1:
    df1['UEMP'+x] = df1['UEMP']*df1[x]
    df1['OBRAT'+x] = df1['OBRAT']*df1[x]
    
y = df1['APPROVE']
# Create new predictor variables
X1 = df1.drop('APPROVE',axis=1)

# Partition the original data set into train (0.8) and test (0.2) data sets
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y, test_size=0.2, random_state=42)

In [48]:
from sklearn.preprocessing import StandardScaler

# Scale the data before fitting and evaluating the model
sc = StandardScaler()
X1_train = sc.fit_transform(X1_train)
X1_test = sc.fit_transform(X1_test)

# Use 5-fold cross-validation on the train data set over a combination of 20 values of lambda and 10 values of alpha
from sklearn.model_selection import KFold
fold = KFold(n_splits=5, random_state=42, shuffle=True)

from sklearn.linear_model import LogisticRegressionCV

searchCV = LogisticRegressionCV(
    Cs=list(np.linspace(0.1,2,20,endpoint=True)) #this corresponds to 1/lambda above
    ,penalty='elasticnet'
    ,l1_ratios=np.linspace(0.1,0.9,10,endpoint=True) #this corresponds to alpha above
    ,scoring='accuracy' #proportion of main diag of confusion matrix
    ,cv=fold
    ,random_state=42
    ,max_iter=10000
    ,fit_intercept=True
    ,solver='saga' #only optimizer available for elasticnet
    ,tol=10
)
logit_cv1 = searchCV.fit(X1_train, y1_train.values.ravel())

# Print the proportion of true predictions for my validation set
proportion1 = logit_cv1.score(X1_test, y1_test)*100
print(f'The proportion of true predictions for my validation set is {round(proportion1,6)} %.')

The proportion of true predictions for my validation set is 89.56743 %.


Using LASSO: 

This time, I want to check if a LASSO regression has a higher proportion than 89.5674%. If it does, I can select nonzero variables from LASSO for my model specification; otherwise I will keep my first model.

In [49]:
# Make a copy of all explanatory variables and remove all interaction terms
df2 = df_new[['APPROVE','HRAT','OBRAT','MHIST','PUB','SELF','CHIST','UEMP','MULTI','COSIGN','MS','LNPR','DEP','SCH','THK','ARAC','ASEX','LOC']].copy()

# Make demeaned variables for each
tmp2 = ['OBRAT','HRAT','OBRAT','MHIST','PUB','SELF','CHIST','UEMP','MULTI','COSIGN','MS','LNPR','DEP','SCH','THK','ARAC','ASEX','LOC']
for x in tmp2:
    df2[x+'_dmean'] = df2[x]-df2[x].mean()

# Make interaction terms for each
tmp2str = []
for x in range(len(tmp2)):
    for y in tmp2[x+1:]:
        if x != y:
            tmp2str.append('('+tmp2[x]+'_dmean:'+y+'_dmean'+')')
            
# Make specifications
f2 = 'APPROVE ~ -1 +' + ''.join([x+'+' for x in tmp2])[:-1] + ' + ' + ''.join([x+'+' for x in tmp2str])[:-1] 

In [50]:
# Create outcome vector and design matrices
y2, X2 = patsy.dmatrices(f2, data=df2, return_type='dataframe')

# Create the indices for the train (80%) and validation (20%) data sets.
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.2, random_state=42)

# Scale the data before fitting and evaluating the model
X2_train = sc.fit_transform(X2_train)
X2_test = sc.fit_transform(X2_test)

In [51]:
# Use Lasso only

lassoCV = LogisticRegressionCV(
    penalty='l1'
    ,scoring='accuracy' #proportion of main diag of confusion matrix
    ,cv=fold
    ,random_state = 42
    ,max_iter=1000000
    ,fit_intercept=True  
    ,solver='saga'
)

logit2_lasso = lassoCV.fit(X2_train, y2_train.values.ravel())


proportionLASSO = round(logit2_lasso.score(X2_test, y2_test),6)*100
print(f'The proportion of true predictions for my validation set using LASSO alone is {proportionLASSO} %.')

The proportion of true predictions for my validation set using LASSO alone is 88.5496 %.


Since the prediction has a lower proportion of accuracy than the first one, I will stop here. Next, I apply Ridge Regression to see if the prediction is more accurate than my first model specification.

Using Ridge:

In [52]:
# Use Ridge alone

ridgeCV = LogisticRegressionCV(
    penalty='l2'
    ,scoring='accuracy' #proportion of main diag of confusion matrix
    ,cv=fold
    ,max_iter=1000000
    ,fit_intercept=True  
    ,solver='saga'
)
logit3_ridge = ridgeCV.fit(X2_train, y2_train.values.ravel())

proportionRidge = round(logit3_ridge.score(X2_test, y2_test),6)*100
print(f'The proportion of true predictions for my validation set using Ridge alone is {proportionRidge} %.')

The proportion of true predictions for my validation set using Ridge alone is 87.5318 %.


As Ridge has a lower proportion than LASSO, I will stop here. My first model obtains the highest proportion, so my alternative model specifiction is:

In [53]:
print(list(X1.columns))
print(f'The coefficients are {logit_cv1.coef_}.')
print(f'The proportion of true predictions based on my best model is {round(proportion1,6)} %.')

['HRAT', 'OBRAT', 'MHIST', 'PUB', 'SELF', 'CHIST', 'UEMP', 'MULTI', 'COSIGN', 'MS', 'LNPR', 'DEP', 'SCH', 'THK', 'ARAC', 'ASEX', 'LOC', 'SCHRAC', 'ACHSEX', 'SCHTHK', 'SCHELOC', 'UEMPARAC', 'OBRATARAC', 'UEMPASEX', 'OBRATASEX', 'UEMPTHK', 'OBRATTHK']
The coefficients are [[ 0.16902022 -0.61264117  0.17605503  0.34487681 -0.26657145  0.53883369
  -0.10191509 -0.08610914  0.00779537  0.13985363 -0.15989656  0.01250593
  -0.16412423  0.00086282  0.35803503 -0.14358337  0.07889224 -0.10060816
   0.09688106  0.07727682  0.13552818  0.01515798  0.10010431 -0.16205141
  -0.15153927  0.21290882 -0.06117876]].
The proportion of true predictions based on my best model is 89.56743 %.
